overlord/devicestate/devicemgr.go

```diff
@@ -378,21 +380,35 @@ func setClassicFallbackModel(st *state.State, device *auth.DeviceState) error {
 	return nil
 }
 
-// SystemMode returns the current mode of the system. Note, that for pre-UC20
-// systems, the mode is not explicitly set and a default "run" mode is always
-// returned.
-func (m *DeviceManager) SystemMode() string {
-	if m.systemMode == "" {
+type SysExpectation int
+
+const (
+	// SysAny indicates any system is appropriate.
+	SysAny SysExpectation = iota
+	// SysHasModeenv indicates only systems with modeenv are appropriate.
+	SysHasModeenv
+)
+
+// SystemMode returns the current mode of the system.
+// An expecation about the system should be passed in, if it's SysAny
+// and the system is pre-UC20 where there is no explicitly mode
+// "run" is returned, if it's SysHasModeenv on such systems it returns
+// simply "".
```

Comment:
```suggestion
// An expectation about the system overrides the returned mode
// when none was set explicitly. In which case, with SysAny, the
// mode defaults to implicit "run", thus covering pre-UC20 systems.
// With SysHasModeeenv, there is no implicit default in systems 
// that use modeenv and thus "" is returned.
```

Reply:
I applied a tweaked version of this that is even more painstaking.